synapse/lib/stormlib/vault.py

```diff
@@ -0,0 +1,649 @@
+        vault = self.runt.snap.core.getVaultByType(vtype, self.runt.user.iden, scope)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        return Vault(self.runt, vault.get('iden'))
+
+    async def _listVaults(self):
+        for vault in self.runt.snap.core.listVaults():
+            if not self.runt.snap.core._hasEasyPerm(vault, self.runt.user, s_cell.PERM_READ):
+                continue
+
+            yield Vault(self.runt, vault.get('iden'))
+
+@s_stormtypes.registry.registerType
+class VaultData(s_stormtypes.Prim):
+    '''
+    Implements the Storm API for Vault data.
+
+    Callers (instantiation) of this class must have already checked that the user has at least
+    PERM_EDIT to the vault.
+    '''
+    _storm_typename = 'vault:data'
+    _ismutable = False
+
+    def __init__(self, runt, valu, path=None):
+        s_stormtypes.Prim.__init__(self, valu, path=path)
+        self.runt = runt
+
+    @s_stormtypes.stormfunc(readonly=False)
+    async def setitem(self, name, valu):
+        vault = self.runt.snap.core.reqVault(self.valu)
+
+        name = await s_stormtypes.tostr(name)
+
+        if valu is s_stormtypes.undef:
+            valu = s_common.novalu
+        else:
+            valu = await s_stormtypes.toprim(valu)
+
+        return await self.runt.snap.core.setVaultData(self.valu, name, valu)
+
+    async def deref(self, name):
+        vault = self.runt.snap.core.reqVault(self.valu)
+
+        name = await s_stormtypes.tostr(name)
+
+        data = vault.get('data')
+        valu = data.get(name, s_common.novalu)
+
+        if valu is not s_common.novalu:
+            return valu
+
+        raise s_exc.NoSuchName(mesg=f'Cannot find name [{name}]', name=name, styp=self.__class__.__name__)
+
+    async def iter(self):
+        vault = self.runt.snap.core.reqVault(self.valu)
+        data = vault.get('data')
+
+        for item in data.items():
+            yield item
+
+    def __len__(self):
+        vault = self.runt.snap.core.reqVault(self.valu)
+        data = vault.get('data')
+        return len(data)
+
+    def stormrepr(self):
+        vault = self.runt.snap.core.reqVault(self.valu)
+        data = vault.get('data')
+        return repr(data)
+
+    def value(self):
+        vault = self.runt.snap.core.reqVault(self.valu)
+        return vault.get('data')
+
+@s_stormtypes.registry.registerType
+class Vault(s_stormtypes.Prim):
+    '''
+    Implements the Storm API for a Vault.
+
+    Callers (instantiation) of this class must have already checked that the user has at least
+    PERM_READ to the vault.
+    '''
+    _storm_locals = (
+        {'name': 'iden', 'desc': 'The Vault iden.', 'type': 'str', },
+        {'name': 'type', 'desc': 'The Vault type.', 'type': 'str', },
+        {'name': 'scope', 'desc': 'The Vault scope.', 'type': 'str', },
+        {'name': 'ident', 'desc': 'The Vault ident (user or role iden).', 'type': 'str', },
+        {'name': 'data', 'desc': 'The Vault data.', 'type': 'vault:data'},
```

Comment:
This should be a `type: { 'type': ['gtor',]...` since we're lazily constructing the object upon access.

Would someone ever want to say `$vault.data = ({"hehe": $superSecretJoke})` ?

Reply:
Oh, your question is about setting the entire dictionary. No, that's not supported. You can pass the entire dictionary with the initial vault definition but after that, you have to use set/get on the data field.